src/modules/pulumi.rs

```diff
@@ -0,0 +1,295 @@
+/// can return results like `3.12.0-alpha.1630554544+f89e9a29.dirty`, which we
+/// don't want to see. Instead we display that as `3.12.0-alpha`.
+fn parse_version(version: &str) -> &str {
+    &version[0..version
+        .trim()
+        .split('.')
+        .take(3)
+        .map(&str::len)
+        .sum::<usize>()
+        + 2]
+}
+
+/// Find a file describing a Pulumi package in the current directory (or any parrent directory).
+fn find_package_file(path: &Path) -> Option<PathBuf> {
+    for path in path.ancestors() {
+        log::trace!("Looking for package file in {:?}", path);
+        let dir = std::fs::read_dir(path).ok()?;
+        let goal = dir.filter_map(Result::ok).find(|path| {
+            path.file_name() == OsStr::new("Pulumi.yaml")
+                || path.file_name() == OsStr::new("Pulumi.yml")
+        });
+        if let Some(goal) = goal {
+            return Some(goal.path());
+        }
+    }
+    log::trace!("Did not find a Pulumi package file");
+    None
+}
+
+/// We get the name of the current stack.
+///
+/// Pulumi has no CLI option that is fast enough to get this for us, but finding
+/// the location is simple. We get it ourselves.
+fn stack_name(project_file: &Path, context: &Context) -> Option<String> {
+    let mut file = File::open(&project_file).ok()?;
+
+    let mut contents = String::new();
+    file.read_to_string(&mut contents).ok()?;
+    let name =
+        YamlLoader::load_from_str(&contents)
+            .ok()
+            .map(|yaml| -> Option<Option<String>> {
+                log::trace!("Parsed {:?} into yaml", project_file);
+                let yaml = yaml.into_iter().next()?;
+                yaml.into_hash().map(|mut hash| -> Option<String> {
+                    hash.remove(&Yaml::String("name".to_string()))?
+                        .into_string()
+                })
+            })???;
+    log::trace!("Found project name: {:?}", name);
+
+    let workspace_file = get_pulumi_workspace(context, &name, project_file)
+        .map(File::open)?
+        .ok()?;
+    log::trace!("Trying to read workspace_file: {:?}", workspace_file);
+    let workspace: serde_json::Value = match serde_json::from_reader(workspace_file) {
+        Ok(k) => k,
+        Err(e) => {
+            log::debug!("Failed to parse workspace file: {}", e);
+            return None;
+        }
+    };
+    log::trace!("Read workspace_file: {:?}", workspace);
+    workspace
+        .as_object()?
+        .get("stack")?
+        .as_str()
+        .map(ToString::to_string)
+}
+
+/// Calculates the path of the workspace settings file for a given pulumi stack.
+fn get_pulumi_workspace(context: &Context, name: &str, project_file: &Path) -> Option<PathBuf> {
+    let project_file = if cfg!(test) {
+        // Because this depends on the absolute path of the file, it changes in
+        // each test run. We thus mock it.
+        "test".to_string()
+    } else {
+        let mut hasher = Sha1::new();
```

Comment:
So pulumi uses a hash of the absolute path as bytes to create a custom workspace dir. What happens if the OS does not encode paths with UTF-8? Would the pulumi tool have to decode, reinterpret as UTF-8, and then re-encode?

(And perhaps the real question: is this anything we're ever going to have to deal with?)

Reply:
It's a good point. Pulumi doesn't deal with any encoding, that is handled by Go's [os.Getwd](https://pkg.go.dev/os#Getwd). As far as I can tell, this translates whatever it gets to utf8, but it's platform specific (In Go, a `string` is just `&[u8]`). I don't think we need to care about invalid utf8. 

If it happens, starship will fail to find the stack name, but will otherwise continue. I don't think it's work caring.